### PR TITLE
MULE-11790: Implement OAuth support in SDK

### PR DESCRIPTION
**Pull request by marianogonzalez**

### CODE REVIEW

File: extensions/email/src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPOperations.java [hunk 1]
```diff
@@ -125,6 +109,22 @@ public void markAsDeleted(@Connection MailboxConnection connection,
...
+   * Marks a single email as READ changing it's state in the specified mailbox folder.
+   * <p>
+   * This operation can targets a single email.
+   *
+   * @param connection The corresponding {@link MailboxConnection} instance.
+   * @param mailboxFolder Folder where the emails are going to be marked as read
+   * @param emailId Email ID Number of the email to mark as read.
+   */
+  @Throws(EmailMarkingErrorTypeProvider.class)
+  public void markAsRead(@Connection MailboxConnection connection,
```
Comment (elrodro83) [inline extensions/email/src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPOperations.java]: unrelated change

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/oauth/BaseOAuthExtensionTestCase.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        .put("state", STATE)
+        .build();
+
+    String localAuthUrl = toUrl(LOCAL_AUTH_PATH, callbackPort.getNumber());
+    Get(localAuthUrl + "?" + encodeQueryString(queryParams))
+        .connectTimeout(REQUEST_TIMEOUT).socketTimeout(REQUEST_TIMEOUT).execute();
+  }
+
+  protected void simulateCallback() {
+    final String authCode = "chu chu ua, chu chu ua";
```
Comment (elrodro83) [inline modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/oauth/BaseOAuthExtensionTestCase.java]: image

File: modules/extensions-support/src/main/resources/META-INF/mule-module.properties [hunk 3]
```diff
@@ -2,6 +2,7 @@ module.name=extensions-support
 
 artifact.export.classPackages=org.mule.runtime.module.extension.internal,\
                               org.mule.runtime.module.extension.internal.config,\
+                              org.mule.runtime.module.extension.internal.runtime.connectivity.oauth,\
```
Comment (elrodro83) [inline modules/extensions-support/src/main/resources/META-INF/mule-module.properties]: surely part of a bigger task, but why are you exporting `internal` stuff?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/DefaultExtensionsOAuthManager.java [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+import java.util.function.Supplier;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Default implementation of {@Link ExtensionsOAuthManager}
+ */
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/DefaultExtensionsOAuthManager.java]: since

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/DefaultExtensionsOAuthManager.java [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+    started = true;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void enable(OAuthConfig config) throws MuleException {
+    final String dancerKey = config.getOwnerConfigName();
+    synchronized (dancers) {
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/DefaultExtensionsOAuthManager.java]: doesn't `ConcurrentHashMap#computeIfAbsent` do all this for you?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthConnectionProviderWrapper.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+                                        Map<Field, String> callbackValues,
+                                        ExtensionsOAuthManager oauthManager,
+                                        boolean disableValidation,
+                                        RetryPolicyTemplate retryPolicyTemplate) {
+    super(delegate, disableValidation, retryPolicyTemplate);
+    this.oauthConfig = oauthConfig;
+    this.oauthManager = oauthManager;
+    authCodeStateSetter = getAuthCodeStateSetter(delegate);
+    this.callbackValues = unmodifiableMap(callbackValues);
+    dance = Once.of(this::updateAuthState);
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthConnectionProviderWrapper.java]: what can fail in `updateAuthState` to put this Once thingy?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthConnectionProviderWrapper.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+  private FieldSetter<ConnectionProvider<C>, AuthorizationCodeState> getAuthCodeStateSetter(ConnectionProvider<C> delegate) {
+    List<Field> stateFields = getFields(delegate.getClass()).stream()
+        .filter(f -> f.getType().equals(AuthorizationCodeState.class))
+        .collect(toList());
+
+    if (stateFields.size() != 1) {
+      throw new IllegalConnectionProviderModelDefinitionException(
+                                                                  format("Connection Provider of class '%s' uses OAuth2 authorization code grant type and thus should contain "
+                                                                      + "one (and only one) field of type %s. %d were found",
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthConnectionProviderWrapper.java]: are we validating this in the connection provider model validators?

File: modules/extensions-support/src/main/resources/META-INF/mule-module.properties (hunk 3 above)
Comment (marianogonzalez) [inline modules/extensions-support/src/main/resources/META-INF/mule-module.properties]: extensions-spring-config requires this

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthConnectionProviderWrapper.java (hunk 6 above)
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthConnectionProviderWrapper.java]: it's not about failing. It's about not injecting the oauth values multiple times

File: core/src/main/java/org/mule/runtime/core/util/store/SimpleObjectStoreToMapAdapter.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.io.Serializable;
+
+/**
+ * Default implementation of {@link ObjectStoreToMapAdapter}
+ *
+ * @param <T> the generic type of the instances contained in the {@link ListableObjectStore}
+ * @since 4.0
+ */
+public class SimpleObjectStoreToMapAdapter<T extends Serializable> extends ObjectStoreToMapAdapter<T> {
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/util/store/SimpleObjectStoreToMapAdapter.java]: It's not default... There's no default... I'll change the jdoc instead

File: core/src/main/java/org/mule/runtime/core/util/func/CheckedRunnable.java [hunk 9]
```diff
@@ -6,19 +6,33 @@
...
+  default void run() {
+    try {
+      runChecked();
+    } catch (Throwable throwable) {
+      handleException(throwable);
+    }
+
+  }
+
+  default void handleException(Throwable throwable) {
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/util/func/CheckedRunnable.java]: done

File: extensions/email/src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPOperations.java [hunk 10]
```diff
@@ -125,6 +109,22 @@ public void markAsDeleted(@Connection MailboxConnection connection,
   }
 
   /**
+   * Marks a single email as READ changing it's state in the specified mailbox folder.
```
Comment (marianogonzalez) [inline extensions/email/src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPOperations.java]: no idea

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/JavaOAuthDeclarationEnricher.java [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        if (!visitedProviders.add(new Reference<>(declaration))) {
+          return;
+        }
+
+        if (declaration.getModelProperty(OAuthModelProperty.class).isPresent()) {
+          if (owner instanceof ExtensionDeclaration) {
+            oauthGloballySupported.set(true);
+            stop();
+          } else if (owner instanceof ConfigurationDeclaration) {
+            oauthConfigs.add(new Reference<>((ConfigurationDeclaration) owner));
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/JavaOAuthDeclarationEnricher.java]: Making sure that the set operates on references and it's not affected by equals and hashCode implementations

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/DefaultExtensionsOAuthManager.java (hunk 5 above)
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/DefaultExtensionsOAuthManager.java]: done

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthConnectionProviderWrapper.java (hunk 7 above)
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthConnectionProviderWrapper.java]: this is an exclusive concern of the Java version of it. Expensive and not all correct to do it in a generic model validator

File: extensions/email/src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPOperations.java (hunk 1 above)
Comment (marianogonzalez) [inline extensions/email/src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPOperations.java]: deal with it

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthConnectionProviderWrapper.java (hunk 6 above)
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthConnectionProviderWrapper.java]: it's not about failing, it's about not doing the same thing on each connect invokation

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/JavaOAuthDeclarationEnricher.java (hunk 11 above)
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/JavaOAuthDeclarationEnricher.java]: but the reference delegates the hash to the value, so it's the same than using the declaration. Anyway, not a big deal, mostly curiosity
